### PR TITLE
Lab homework for tmarsha1

### PR DESCRIPTION
**Pull request by TroyLMarshall**

### CODE REVIEW

File: training/level-2-command-line-interfaces/dragon-warrior/tmarsha1/file_ops.py [hunk 1]
```diff
@@ -1,21 +1,125 @@
...
+Functions:
+    copy_files: copy file(s) to a specific location.
+    move_files: move file(s) to a specific location.
+"""
 
 import subprocess
 
 
-def copy_files(files: list, destination: str):
+class FolderNotFoundError(Exception):
```
Comment (eikonomega) [inline training/level-2-command-line-interfaces/dragon-warrior/tmarsha1/file_ops.py:12]: @TroyLMarshall Excellent.  Custom exceptions.  It is common to see exceptions defined in a separate module.  Something to consider.

File: training/level-2-command-line-interfaces/dragon-warrior/tmarsha1/file_ops.py [hunk 2]
```diff
@@ -1,21 +1,125 @@
...
+    def __str__(self):
+        return repr(self.value)
+
+
+def _process_files(
+        command: str, 
+        options: str, 
+        file: str, 
+        destination: str
+) -> str:
```
Comment (eikonomega) [inline training/level-2-command-line-interfaces/dragon-warrior/tmarsha1/file_ops.py:31]: @TroyLMarshall Could just be Github, but it looks like the function parameters in the definition are double indented.  Also, standard style is not to put the closing parenthesis on a separate line.

File: training/level-2-command-line-interfaces/dragon-warrior/tmarsha1/file_ops.py [hunk 3]
```diff
@@ -1,21 +1,125 @@
...
+            raise FileNotFoundError()
+        elif "Not a directory" in message:
+            raise FolderNotFoundError(destination)
+        elif "Permission" in message:
+            raise InsufficientRightsError(destination)
+        else:
+            print(message)
+    else:
+        result = byte_result.decode()
+        result = result. rstrip()
```
Comment (eikonomega) [inline training/level-2-command-line-interfaces/dragon-warrior/tmarsha1/file_ops.py:49]: @TroyLMarshall Ok if you are going for clarity, but not necessary. Could just be: `return byte_result.decode().rstrip().  Are you gaining something from putting it in three steps?

Judgement call, but seems a little more than necessary.

File: training/level-2-command-line-interfaces/dragon-warrior/tmarsha1/git_cli.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
+"""
+This module provides a CLI that copies files using the 
+Click third-party library.
+"""
+
+import click
+
+import git_ops
```
Comment (eikonomega) [inline training/level-2-command-line-interfaces/dragon-warrior/tmarsha1/git_cli.py:8]: Good job remembering PEP8 syntax rules for imports.

File: training/level-2-command-line-interfaces/dragon-warrior/tmarsha1/git_ops.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+
+def git_fetch(remote: str):
+    """
+    Get files from a specific remote.
+
+    Args:
+        remote: A string specifying the remote host.
+    """
+
```
Comment (eikonomega) [inline training/level-2-command-line-interfaces/dragon-warrior/tmarsha1/git_ops.py:19]: This space is unnecessary, but ok if desired.

**Comment by eikonomega** [comment]
@TroyLMarshall Let me know that you've seem my comments and I'll merge.

**Comment by TroyLMarshall** [comment]
Yes, I would normally put the custom exceptions into a separate module. I
was more interested in using the exceptions and focusing on getting some of
the other functionality working.

My guess with the double indent is that it's probably related to me writing
the code in Notepad++ instead of PyCharm. Fortunately, I received my
license on Monday and have applied it to my copy already.

I'm not a big fan of chaining calls. In this particular case, it's probably
ok, but chaining calls like that can introduced hidden dependencies that
are difficult to maintain long term.

Troy

**Comment by eikonomega** [comment]
Cool.

On Wed, Nov 11, 2015 at 9:32 AM TroyLMarshall notifications@github.com
wrote:

> Yes, I would normally put the custom exceptions into a separate module. I
> was more interested in using the exceptions and focusing on getting some of
> the other functionality working.
> 
> My guess with the double indent is that it's probably related to me writing
> the code in Notepad++ instead of PyCharm. Fortunately, I received my
> license on Monday and have applied it to my copy already.
> 
> I'm not a big fan of chaining calls. In this particular case, it's probably
> ok, but chaining calls like that can introduced hidden dependencies that
> are difficult to maintain long term.
> 
> Troy
> 
> On Wed, Nov 11, 2015 at 9:03 AM, Michael Dunn notifications@github.com
> wrote:
> 
> > @TroyLMarshall https://github.com/TroyLMarshall Let me know that
> > you've
> > seem my comments and I'll merge.
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > <
> > https://github.com/bigfatpanda-training/pandas-practical-python-primer/pull/59#issuecomment-155791378
> > 
> > .
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/bigfatpanda-training/pandas-practical-python-primer/pull/59#issuecomment-155798000
> .
> 
> ## 
> 
> ---
> 
> Mike Dunn
> https://about.me/eikonomega

Returning hate for hate multiplies hate, adding deeper darkness to a night
already devoid of stars. Darkness cannot drive out darkness; only light can
do that. Hate cannot drive out hate; only love can do that. - MLK

**Pull request merged**